cowsay.1
'\" t
.\"     Title: cowsay
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.23
.\"      Date: 2024-08-16
.\"    Manual: Cowsay Manual
.\"    Source: Cowsay
.\"  Language: English
.\"
.TH "COWSAY" "1" "2024-08-16" "Cowsay" "Cowsay Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
cowsay, cowthink \- configurable speaking/thinking cow (and a bit more)
.SH "SYNOPSIS"
.sp
\fBcowsay\fP
[\fB\-e\fP \fIeye_string\fP]
[\fB\-f\fP \fIcowfile\fP | \fB\-r\fP [\fB\-C\fP] ]
[\fB\-h\fP]
[\fB\-l\fP]
[\fB\-n\fP]
[\fB\-T\fP \fItongue_string\fP]
[\fB\-W\fP \fIcolumn\fP]
[\fB\-bdgpstwy\fP]
[<message>]
.SH "DESCRIPTION"
.sp
\fICowsay\fP generates an ASCII picture of a cow saying something provided by the user.  If run with no arguments, it accepts standard input, word\-wraps the message given at about 40 columns, and prints the cow saying the given message on standard output.
.sp
To aid in the use of arbitrary messages with arbitrary whitespace, use the \fB\-n\fP option.  If it is specified, the given message will not be word\-wrapped.  This is possibly useful if you want to make the cow think or speak in figlet(6).  If \fB\-n\fP is specified, there must not be any command line arguments left after all the switches have been processed.
.sp
The \fB\-W\fP specifies roughly where the message should be wrapped. The default is equivalent to \fB\-W 40\fP (i.e. wrap words at or before the 40th column).
.sp
If any command\-line arguments are left over after all switches have been processed, they become the cow\(cqs message. The program will not accept standard input for a message in this case.
.sp
If the program is invoked as \fBcowthink\fP then the cow will think its message instead of saying it.
.SH "OPTIONS"
.sp
There are several provided modes which change the appearance of the cow depending on its particular emotional/physical state.
.SS "Cow modification"
.sp
\fB\-b\fP
.RS 4
Invokes Borg mode.
.RE
.sp
\fB\-d\fP
.RS 4
Causes the cow to appear dead.
.RE
.sp
\fB\-g\fP
.RS 4
Invokes greedy mode.
.RE
.sp
\fB\-p\fP
.RS 4
Causes a state of paranoia to come over the cow.
.RE
.sp
\fB\-s\fP
.RS 4
Makes the cow appear thoroughly stoned.
.RE
.sp
\fB\-t\fP
.RS 4
Yields a tired cow.
.RE
.sp
\fB\-w\fP
.RS 4
Is somewhat the opposite of \fB\-t\fP, and initiates wired mode.
.RE
.sp
\fB\-y\fP
.RS 4
Brings on the cow\(cqs youthful appearance.
.RE
.sp
\fB\-e\fP
.RS 4
Selects the appearance of the cow\(cqs eyes, in which case the first two characters of the argument string \fIeye_string\fP will be used.  The default eyes are \fBoo\fP.  The tongue is similarly configurable through \fB\-T\fP and \fItongue_string\fP; it must be two characters and does not appear by default.  However, it does appear in the \fIdead\fP and \fIstoned\fP modes.  Any configuration done by \fB\-e\fP and \fB\-T\fP will be lost if one of the provided modes is used.
.RE
.SS "Cow selection"
.sp
\fB\-f\fP
.RS 4
Specifies a particular cow picture file (\fIcowfile\fP) to use.  If the cowfile spec resolves to an existing file, then it will be interpreted as a path to the cowfile.  Otherwise, cowsay will search the COWPATH for a cowfile with that name. Additional cowpath entries may be specified in the \fBCOWPATH\fP environment variable. To list all cowfiles on the current \fBCOWPATH\fP, invoke \fBcowsay\fP with the \fB\-l\fP switch.
.RE
.sp
\fB\-r\fP
.RS 4
Chooses a random cow from the cows on the \fBCOWPATH\fP.
.RE
.sp
\fB\-C\fP
.RS 4
Enables true\-color (24\-bit) cows. Currently only has an effect when used in conjunction with \fB\-r\fP.
.RE
.SS "Non\-cow operations"
.sp
\fB\-l\fP
.RS 4
Lists the defined cows on the current \fBCOWPATH\fP. Displays it in a human\-readable pretty\-printed format when displaying to a terminal device. When sent to a non\-terminal device, outputs the list in a parsing\-friendly format with one cow name per line and no headers or blank lines.
.RE
.sp
\fB\-h\fP
.RS 4
Displays a help screen.
.RE
.SH "COWFILE FORMAT"
.sp
A cowfile is made up of a simple block of \fBperl(1)\fP code, which assigns a picture of a cow to the variable \fB$the_cow\fP. Should you wish to customize the eyes or the tongue of the cow, then the variables \fB$eyes\fP and \fB$tongue\fP may be used.  The trail leading up to the cow\(cqs message balloon is composed of the character(s) in the \fB$thoughts\fP variable.  Any backslashes must be reduplicated to prevent interpretation.  Also, at\-signs ("@") must be backslashed because that is what Perl 5 expects. The name of a cowfile must end with \fB.cow\fP, otherwise it is not recognized as a cowfile.
.SH "ENVIRONMENT"
.sp
The \fBCOWPATH\fP environment variable, if present, will be used to search for cowfiles.  It contains a colon\-separated list of directories, much like \fBPATH\fP or \fBMANPATH\fP.
.sp
The default \fBCOWPATH\fP is searched after entries in the user\-specified \fBCOWPATH\fP environment variable.  To suppress this behavior (e.g. for development work), set the \fBCOWSAY_ONLY_COWPATH\fP environment variable to \fB1\fP. In this case, \fBCOWPATH\fP should contain at least a directory with a file called \fBdefault.cow\fP in it.
.SH "FILES"
.sp
\fB%PREFIX%/share/cowsay/cows\fP holds a sample set of cowfiles.  Your \fBCOWPATH\fP automatically contains this directory, unless \fBCOWSAY_ONLY_COWPATH\fP is set to \fB1\fP.
.sp
\fB%PREFIX%/share/cowsay/site\-cows\fP is provided for administrators to install custom cows. Cows in \fBshare/cowsay/site\-cows\fP take precedence over cows with the same name in \fBshare/cowsay/cows\fP. The \fBsite\-cows\fP directory will never be modified by cowsay installations, so custom cows defined there will persist across upgrades of cowsay. \fBsite\-cows\fP is also on the default \fBCOWPATH\fP.
.sp
\fB%PREFIX%/etc/cowsay/cowpath.d/\fP (or \fB/etc/cowsay/cowpath.d/\fP when %PREFIX% is \fB/usr\fP) is a directory that contains files which list entries to be added to the default \fBCOWPATH\fP. This mechanism allows third\-party cow collections to register themselves with cowsay in a way that does not require per\-user configuration.
.SH "BUGS"
.sp
If there are any, please report them on the cowsay GitHub page (\c
.URL "https://github.com/cowsay\-org/cowsay/issues" "" ")"
or notify the author at the address below.
.SH "AUTHOR"
.sp
Cowsay is maintained by Andrew Janke (\c
.MTO "floss\(atapjanke.net" "" ")."
.sp
Cowsay was originally written by Tony Monroe (\c
.MTO "tony\(atnog.net" "" "),"
with suggestions from Shannon Appel (
.MTO "appel\(atcsua.berkeley.edu" "" ")"
and contributions from Anthony Polito (
.MTO "aspolito\(atcsua.berkeley.edu" "" ")."
.SH "RESOURCES"
.sp
GitHub: \c
.URL "https://github.com/cowsay\-org/cowsay" "" ""
.sp
Main web site: \c
.URL "http://cowsay.diamonds" "" ""
.SH "SEE ALSO"
.sp
\fBfortune(1)\fP, \fBperl(1)\fP, \fBwall(1)\fP, \fBnwrite(1)\fP, \fBfiglet(6)\fP

ChangeLog
Cowsay Changelog
================

3.8.0 (in progress)

- Add "cupcake", "alpaca", and "fox" cows
- Fix "mech-and-cow" cow for file extension issue
- Better installation stuff, involving symlinks and stuff
- Make cowthink a symlink
- Improve man page organization

3.7.0 (2020-05-28)

- `cowsay -r [-C]` for random cow
- Pluggable cow collections using etc/cowsay/cowpath.d
- Move cows/ and site-cows/ from $prefix/share/ to $prefix/share/cowsay/
- Parseable format for `cowsay -l` when sent to non-terminal destination
- Fix `cowsay -l` ignoring `*.pm` files
- Have `cowsay -l` recursively discover cowfiles in subdirectories

3.6.1 (2020-05-28)

- Fix version number

3.6.0 (2020-05-28)

- Support running from repo
- Simplify installation process

3.5.0 (2020-05-28)

- Add llama cow
- Changed to SemVer version numbering
- Remove leading zeros from version numbering

3.04.01

- Cowsay forked by Andrew Janke.
- Fixed version numbers in program and doco

-- Tony Monroe's original cowsay from here on down --

3.04 24 June 2016

- Licensing terms have changed to GPLv3.
- Friendlier to being hosted on GitHub.
- No functionality changes whatsoever.

3.03 28 May 1999

- Added cows/tux.cow, as suggested by xmanoel@i.am
- Compatibility with 5.6.0, due to a change in qw().
- Renamed devil.cow to daemon.cow, since I know better. :-)

3.02 04 November 1999

- Fixed boneheaded code placement so that cowsay -l actually works.

3.01 01 November 1999

- Fixed compatibility issues between the Text::Wrap module that changed between 5.005_02 and 5.005_03.
- Fixed tab expansion issues with Text::Tabs.

3.0 13 April 1999, released 14 August 1999

- Rewritten into Perl 5 and presented to the world.

-- Not present in CVS from here on down --

2.x Date?

- Arbitrary messages. 
- Figlet support (-n). 
- Line wrap length (-w). 
- Multiple pre-set expressions.
- Better arg parsing loop.
- Message from stdin or command line.

1.0 Date?

- SUBJECT is VERB OBJECT


cowsay.pl
#%BANGPERL%

##
## Cowsay
##

use Text::Tabs qw(expand);
use Text::Wrap qw(wrap fill $columns);
use File::Basename;
use Getopt::Std;
use Cwd;

$version = "3.6.0-SNAPSHOT";
$progname = basename($0);
$eyes = "oo";
$tongue = "  ";
$pathsep = detect_pathsep();
$default_cowpath = join($pathsep, '%PREFIX%/share/site-cows', '%PREFIX%/share/cows');
$cowpath = pick_cowpath();
@cowpath = split(/$pathsep/, $cowpath);
@message = ();
$thoughts = "";

## Yeah, this is rude, I know.  But hopefully it gets around a nasty
## little version dependency.

$Text::Wrap::initial_tab = 8;
$Text::Wrap::subsequent_tab = 8;
$Text::Wrap::tabstop = 8;

%opts = (
    'e'         =>      'oo',
    'f'         =>      'default.cow',
    'n'         =>      0,
    'T'         =>      '  ',
    'W'         =>      40,
);

getopts('bde:f:ghlLnNpstT:wW:y', \%opts);

&display_usage if $opts{'h'};
&list_cowfiles if $opts{'l'};

$borg = $opts{'b'};
$dead = $opts{'d'};
$greedy = $opts{'g'};
$paranoid = $opts{'p'};
$stoned = $opts{'s'};
$tired = $opts{'t'};
$wired = $opts{'w'};
$young = $opts{'y'};
$eyes = substr($opts{'e'}, 0, 2);
$tongue = substr($opts{'T'}, 0, 2);
$the_cow = "";

&slurp_input;
$Text::Wrap::columns = $opts{'W'};
@message = ($opts{'n'} ? expand(@message) : 
            split("\n", fill("", "", @message)));
&construct_balloon;
&construct_face;
&get_cow;
print @balloon_lines;
print $the_cow;

sub detect_pathsep {
    ## One of these days, we'll get it ported to Windows.  Yeah, right.
    if (($^O eq "MSWin32") or ($^O eq "Windows_NT")) {  ## Many perls, eek!
        return ';';
    } else {
        return ':';
    }
}

sub pick_cowpath {
    my $cowpath;
    if ($ENV{'COWPATH'}) {
        if ($ENV{'COWSAY_ONLY_COWPATH'} == 1) {
            $cowpath = $ENV{'COWPATH'}
        } else {
            $cowpath = $ENV{'COWPATH'} + $pathsep + $default_cowpath
        }
    } else {
        $cowpath = $default_cowpath;
    }
    return $cowpath;
}

sub list_cowfiles {
    my $basedir;
    my @dirfiles;
    chop($basedir = cwd);
    for my $d (split(/$pathsep/, $cowpath)) {
        print "Cow files in $d:\n";
        opendir(COWDIR, $d) || die "$0: Cannot open $d\n";
        for my $file (readdir COWDIR) {
            if ($file =~ s/\.cow$//) {
                push(@dirfiles, $file);
            }
        }
        closedir(COWDIR);
        print wrap("", "", sort @dirfiles), "\n";
        @dirfiles = ();
        chdir($basedir);
    }
    exit(0);
}

sub slurp_input {
    unless ($ARGV[0]) {
        chomp(@message = <STDIN>);
    } else {
        &display_usage if $opts{'n'};
        @message = join(' ', @ARGV);
    }
}

sub maxlength {
    my ($l, $m);
    $m = -1;
    for my $i (@_) {
        $l = length $i;
        $m = $l if ($l > $m);
    }
    return $m;
}

sub construct_balloon {
    my $max = &maxlength(@message);
    my $max2 = $max + 2;        ## border space fudge.
    my $format = "%s %-${max}s %s\n";
    my @border; ## up-left, up-right, down-left, down-right, left, right
    if ($0 =~ /think/i) {
        $thoughts = 'o';
        @border = qw[ ( ) ( ) ( ) ];
    } elsif (@message < 2) {
        $thoughts = '\\';
        @border = qw[ < > ];
    } else {
        $thoughts = '\\';
        if ($V and $V gt v5.6.0) {              # Thanks, perldelta.
            @border = qw[ / \\ \\ / | | ];
        } else {
            @border = qw[ / \ \ / | | ];        
        }
    }
    push(@balloon_lines, 
        " " . ("_" x $max2) . " \n" ,
        sprintf($format, $border[0], $message[0], $border[1]),
        (@message < 2 ? "" : 
            map { sprintf($format, $border[4], $_, $border[5]) } 
                @message[1 .. $#message - 1]),
        (@message < 2 ? "" : 
            sprintf($format, $border[2], $message[$#message], $border[3])),
        " " . ("-" x $max2) . " \n"
    );
}

sub construct_face {
    if ($borg) { $eyes = "=="; }
    if ($dead) { $eyes = "xx"; $tongue = "U "; }
    if ($greedy) { $eyes = "\$\$"; }
    if ($paranoid) { $eyes = "@@"; }
    if ($stoned) { $eyes = "**"; $tongue = "U "; }
    if ($tired) { $eyes = "--"; } 
    if ($wired) { $eyes = "OO"; } 
    if ($young) { $eyes = ".."; }
}

sub get_cow {
##
## Get a cow from the specified cowfile; otherwise use the default cow
## which was defined above in $the_cow.
##
    my $f = $opts{'f'};
    my $full = "";
    if ($opts{'f'} =~ m,/,) {
        $full = $opts{'f'};
    } else {
        for my $d (split(/:/, $cowpath)) {
            if (-f "$d/$f") {
                $full = "$d/$f";
                last;
            } elsif (-f "$d/$f.cow") {
                $full = "$d/$f.cow";
                last;
            }
        }
        if ($full eq "") {
            die "$progname: Could not find $f cowfile!\n";
        }
    }
    do $full;
    die "$progname: $@\n" if $@;
}

sub display_usage {
        die <<EOF;
cow{say,think} version $version
Usage: $progname [-bdgpstwy] [-h] [-e eyes] [-f cowfile] 
          [-l] [-n] [-T tongue] [-W wrapcolumn] [message]
EOF
}
